### PR TITLE
Fix `torch.nn.utils.rnn.unpad_sequence` modifies `padded_sequences` in-place

### PR DESCRIPTION
**Pull request by yuanx749**
Fixes #98499
test added

### CODE REVIEW

File: torch/nn/utils/rnn.py [hunk 1]
```diff
@@ -435,7 +435,7 @@ def unpad_sequence(
     unpadded_sequences = []
 
     if not batch_first:
-        padded_sequences.transpose_(0, 1)
+        padded_sequences = padded_sequences.transpose(0, 1)
```
Comment (eqy) [inline torch/nn/utils/rnn.py:438]: Would it make sense to make the in-place behavior optional (even if off by default) as I assume this incurs some memory overhead?

Comment (yuanx749) [reply torch/nn/utils/rnn.py]: Right. I added an ``inplace`` argument.

**Comment by yuanx749** [comment]
Gentle ping for review. @jbschlosser @mikaylagawarecki

File: torch/nn/utils/rnn.py [hunk 2]
```diff
@@ -428,14 +436,19 @@ def unpad_sequence(
...
+            into ``B x T x *`` format in-place when ``batch_first`` is ``False``. Default: ``False``.
 
     Returns:
         a list of :class:`Tensor` objects
     """
     unpadded_sequences = []
 
     if not batch_first:
-        padded_sequences.transpose_(0, 1)
+        if inplace:
```
Comment (mikaylagawarecki) [inline torch/nn/utils/rnn.py:448]: If we are preserving the current behavior (not breaking BC), I think `inplace` should be `True` by default.

**Review by mikaylagawarecki — COMMENTED** [review]
Apologies for the delay @yuanx749 , will get back to you on whether (1) adding a flag makes sense here or (2) the existing behavior should just be documented or (3) `transpose_` --> `transpose` can count as a bugfix

That said, my understanding is that inplace transpose would modify the strides metadata inplace while out of place transpose would create a new `TensorImpl` with the metadata for strides updated but does not modify the storage, so not sure how significant the overhead would be, though I might definitely have missed something @eqy

**Pull request closed without merging**